### PR TITLE
WIP: Structure Python bindings into Python package

### PR DESCRIPTION
**Pull request by robbmcleod**
Nowhere near complete yet, I just want a place to discuss how to go about integrating the CMake build process, the Appveyor/TravisCI integration services, and Python packages.

- Copy the structure in `Components/Python` into the build directory provided 
  by CMake.
- Have CMake set environment variables that can be passed into `setup.py`.
- Ensure that `setup.py` has all the files it needs, including possibly 
  default media files?

### CODE REVIEW

File: Components/Python/BUILD_WIN.txt [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+`OGRE_BUILD_COMPONENT_PYTHON` is checked. 
+
+IMPORTANT: at present (2018-Dec-30), you will also have to turn off
+`OGRE_BUILD_COMPONENT_CSHARP` bindings in the configuration step, as they are 
+broken.
+
+`Configure` again, then `Generate` and finally `Open Project` in MSVC2017. 
+Then you can build the entire solution, which takes about five minutes.
+
+Co-locating the Components
```
Comment (paroj) [inline Components/Python/BUILD_WIN.txt]: if you build the install target all of them are in `build/sdk/bin`

File: Components/Python/BUILD_WIN.txt [hunk 2]
```diff
@@ -0,0 +1,208 @@
+Python for Ogre on Windows
```
Comment (paroj) [inline Components/Python/BUILD_WIN.txt]: probably this can ultimately be folded into BuildingOgre.md

File: appveyor.yml [hunk 3]
```diff
@@ -2,34 +2,67 @@ platform: x64
...
+
+init:
+  - "ECHO %PYTHON_VERSION% %MINICONDA%"
+
+install:
+ - "set PATH=%MINICONDA%;%MINICONDA%\\Scripts;%PATH%"
+ - "python --version"
+ - conda update -q conda
+ - conda info -a
+ - conda create -q -n env_ogre python=%PYTHON_VERSION% numpy freetype freeimage swig
```
Comment (paroj) [inline appveyor.yml]: freeimage is optional. I would probably rather go without it by default

File: appveyor.yml [hunk 4]
```diff
@@ -2,34 +2,67 @@ platform: x64
 configuration: Release
 image: 
  - Visual Studio 2017
- - Visual Studio 2013
+#  - Visual Studio 2013
+
+# Do we want to have a Matrix? 32-bit and 64-bit?
```
Comment (paroj) [inline appveyor.yml]: 64bit only should be enough. We only provide a 64bit SDK and nobody complained so far

File: Samples/Python/ex_pyside2.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    A work-in-progress example of how to embed Ogre into a PySide2/Qt5 application.
+    """
+
+    def __init__(self, parent: qw.QWidget=None):
+        qw.QOpenGLWidget.__init__(self, parent=parent)
+        self._root = Ogre.Root()
+        
+        # Ensure we are using the basic OpenGL renderer
+        for renderer in self._root.getAvailableRenderers():
+            if renderer.getName() == 'OpenGL Rendering Subsystem':
```
Comment (paroj) [inline Samples/Python/ex_pyside2.py]: "OpenGL" in renderer.getName()

**Comment by robbmcleod** [comment]
Is there any nice way to build Appveyor interactively, instead of making one tiny incremental commit after another? I tried searching online but didn't find anything.

**Comment by paroj** [comment]
this maybe? 
https://www.appveyor.com/docs/how-to/rdp-to-build-worker/

**Comment by paroj** [comment]
we now have working python bindings in master via #1069

**Comment by robbmcleod** [comment]
Ok, managed to login via RDP, waiting to see the result of the build for the 2017 build.

Will probably have to figure out how to disable the Python build for MSVC2013, if continued support of that compiler is important, as I believe there is no Python version that supports it:

https://wiki.python.org/moin/WindowsCompilers

**Comment by robbmcleod** [comment]
The RDP session dies as soon as the build errors, so not very useful...

I'm back to building locally, trying to use the `chocolatey` version of `swig`, which seems to be working a bit better than the `conda` version. In which case we should consider building with the C:\Python37 install, and use `pip` to install `numpy` instead.

**Comment by robbmcleod** [comment]
A local environment can mimic Appveyor by setting:

    set APPVEYOR=1
    set APPVEYOR_BUILD_WORKER_IMAGE="Visual Studio 2017"

The `ci-build.cmake` script seems to search for SWIG in different places. Rather than specifying paths it's probably better to just call `find_package(SWIG 3.0.8)` although this needs to be checked on Linux as well. On my system, the `.py` files end up in `ogre\build\Components\Python` and the rest of the library in `ogre\build\lib\Release` including the `.pyd` files, so some copying to a central location, along with the `ogre\Components\Python` package itself will have to be done within the `ogre\Components\Python\CMakeLists.txt` script I suspect.

**Comment by paroj** [comment]
> so some copying to a central location

this happens if you build the INSTALL target. Although the locations are currently derived from where Linux would expect them.

**Comment by robbmcleod** [comment]
I am building the INSTALL target. I don't know where it installs though. Looking at the generated Visual Studio projects there's no commands to copy files.

Is it supposed to be this line?

    set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/${OGRE_LIB_DIRECTORY}")

**Comment by paroj** [comment]
these lines:
https://github.com/OGRECave/ogre/blob/master/Components/Python/CMakeLists.txt#L25-L26

on windows stuff ends up in build/sdk/ by default

**Comment by paroj** [comment]
closing as there is now a conda-forge project: https://github.com/conda-forge/ogre-feedstock

**Pull request closed without merging**